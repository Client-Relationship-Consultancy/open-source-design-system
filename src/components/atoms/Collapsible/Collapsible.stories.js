import React from "react"

import { storiesOf } from "@storybook/react"
import { boolean } from "@storybook/addon-knobs"

import { Collapsible } from "../../../index"

const stories = storiesOf("Collapsible", module)

const collapsedContent = (
  <>
    <h1>Panel</h1>
    <p>This is some text to help explain what this panel is.</p>
    <p>
      <em>Click the arrow to show the hidden content...</em>
    </p>
  </>
)

const openContent = (
  <>
    <p>This is the hidden content you seek!</p>
    <p>This is the hidden content you seek!</p>
    <p>This is the hidden content you seek!</p>
    <p>This is the hidden content you seek!</p>
    <p>This is the hidden content you seek!</p>
  </>
)

stories.add("Collapsible", () => {
  const open = boolean("Open", false)
  const locked = boolean("Locked", false)

  return (
    <Collapsible
      open={open}
      locked={locked}
      collapsedContent={collapsedContent}
      clickHandler={() => console.log("clicked!")}
    >
      {openContent}
    </Collapsible>
  )
})
