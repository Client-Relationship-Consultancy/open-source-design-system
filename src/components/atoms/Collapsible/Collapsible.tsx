import React from "react";
import styled from "styled-components";

import Icon from "../Icon";

const StyledButton = styled.button`
  position: absolute;
  top: 0;
  right: 0;
  background-color: transparent;
  border: 0;
  outline: none;
  cursor: pointer;
`;

const CollapsiblePanel = styled.div`
  position: relative;
  font-size: 1rem;
  min-height: 2em;
  width: 100%;
`;

CollapsiblePanel.displayName = "CollapsiblePanel";

interface ICollapsibleIconProps {
  open: boolean;
}

const CollapsibleIcon = styled(Icon)<ICollapsibleIconProps>`
  svg {
    transform: ${({ open }) => (open ? "rotateZ(180deg)" : "rotateZ(0deg)")};
    transition: transform 0.5s;
  }
`;

const CollapsedContentContainer = styled.div``;

CollapsedContentContainer.displayName = "CollapsedContentContainer";

interface IOpenContentContainerProps {
  open: boolean;
}

const OpenContentContainer = styled.div<IOpenContentContainerProps>`
  visibility: ${({ open }) => (open ? "visible" : "hidden")};
  height: ${({ open }) => (open ? "100%" : "0%")};
  opacity: ${({ open }) => (open ? 1 : 0)};
  transition-duration: 0s, 1s, 1s;
  transition-delay: ${({ open }) => (open ? "0s" : "1s")}, 0s, 0s;
  transition-property: visibility, height, opacity;
`;

OpenContentContainer.displayName = "OpenContentContainer";

interface ICollapsibleProps {
  open: boolean;
  locked?: boolean;
  children: React.ReactNode;
  collapsedContent?: React.ReactNode;
  width?: number;
  index: number;
  clickHandler: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

const Collapsible: React.FC<ICollapsibleProps> = (props) => {
  const { locked = false, open, children, collapsedContent, clickHandler, index } = props;

  return (
    <CollapsiblePanel>
      <StyledButton type="button" data-index={index} onClick={clickHandler}>
        <CollapsibleIcon name="chevron-down" size="1x" open={open} />
      </StyledButton>
      {collapsedContent && (
        <CollapsedContentContainer>{collapsedContent}</CollapsedContentContainer>
      )}
      <OpenContentContainer open={open}>{children}</OpenContentContainer>
    </CollapsiblePanel>
  );
};

Collapsible.displayName = "Collapsible";

export default Collapsible;
