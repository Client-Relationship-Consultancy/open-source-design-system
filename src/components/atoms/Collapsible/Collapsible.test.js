import React from "react"
import { shallow } from "enzyme"
import Collapsible from "./Collapsible"

describe("Collapsible Component Test", () => {
  const mockCollapsedContent = (
    <>
      <h1>Mock Panel Header</h1>
      <p>Mock text that explains what this panel does</p>
    </>
  )

  const mockOpenContent = (
    <>
      <p>Mock text that is shown/hidden</p>
      <p>Mock text that is shown/hidden</p>
      <p>Mock text that is shown/hidden</p>
      <p>Mock text that is shown/hidden</p>
      <p>Mock text that is shown/hidden</p>
    </>
  )

  const component = shallow(
    <Collapsible
      open
      locked={false}
      collapsedContent={mockCollapsedContent}
      clickHandler={() => null}
    >
      {mockOpenContent}
    </Collapsible>,
  )

  it("Match last snapshot", () => {
    expect(component).toMatchSnapshot()
  })
})
