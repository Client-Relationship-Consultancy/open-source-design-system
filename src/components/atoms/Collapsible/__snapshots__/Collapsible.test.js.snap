// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible Component Test Match last snapshot 1`] = `
<CollapsiblePanel
  locked={false}
  open={true}
>
  <Collapsible__CollapsibleIcon
    name="chevron-down"
    onClick={[Function]}
    open={true}
    size="1x"
  />
  <CollapsedContentContainer>
    <h1>
      Mock Panel Header
    </h1>
    <p>
      Mock text that explains what this panel does
    </p>
  </CollapsedContentContainer>
  <OpenContentContainer
    open={true}
  >
    <p>
      Mock text that is shown/hidden
    </p>
    <p>
      Mock text that is shown/hidden
    </p>
    <p>
      Mock text that is shown/hidden
    </p>
    <p>
      Mock text that is shown/hidden
    </p>
    <p>
      Mock text that is shown/hidden
    </p>
  </OpenContentContainer>
</CollapsiblePanel>
`;
